src/object/srv_obj.c

```diff
@@ -5051,8 +5083,27 @@ ds_obj_cpd_handler(crt_rpc_t *rpc)
 	D_FREE(dcas);
 	if (dcbs != NULL) {
 		for (i = 0; i < dcb_nr; i++) {
-			if (dcbs[i] != NULL)
-				D_FREE(dcbs[i]->dcb_iov.iov_buf);
+			if (dcbs[i] == NULL)
+				continue;
+
+			if (dcbs[i]->dcb_reqs != NULL) {
+				D_ASSERT(dcbs[i]->dcb_proc != NULL);
+
+				crt_proc_reset(dcbs[i]->dcb_proc, dcbs[i]->dcb_iov.iov_buf,
+					       dcbs[i]->dcb_iov.iov_len, CRT_PROC_FREE);
+				for (j = 0; j < dcbs[i]->dcb_item_nr; j++) {
+					crt_proc_struct_daos_cpd_sub_req(dcbs[i]->dcb_proc,
+									 CRT_PROC_FREE,
+									 &dcbs[i]->dcb_reqs[j],
+									 true);
```

Comment:
may need to confirm all allocated buffers are freed by the added proc by adding special logs in test, just FYI.

Reply:
I can confirm that the DRAM allocated via the CPD RPC proc interface are freed, but I do not know how to verify whether lower layer (CaRT and mercury) allocated buffer are free or not.